### ISSUE TITLE
Problem: PUB/SUB lost the message If I send consecutive packet without sleep

### ISSUE THREAD

**Issue by clearday4** [zeromq/libzmq#2291]
I was measuring the throughput and latency with zeromq pub/sub socket.
Testing with single publisher and subscriber.
Publisher send more than 10000 packets and each packet size is 1024 byte.
And I applied the zerocopy option to the publisher, but the symptoms are different before applying it.
If I apply the zerocopy option to the publisher, subscriber will receive the last message over and over again.
And sub can ont receive all packets. 
All problems work properly if I add sleep() after sending 100 packets.

here is the log if I send 10000 packets. 
packet no. inscreases sequentially by every packet.
but sub receive the same message repeatedly. 

...........
received 1024 packet, packet no=9500 sendTime=1483527718432518
received 1024 packet, packet no=9500 sendTime=1483527718432518
received 1024 packet, packet no=9500 sendTime=1483527718432518
received 1024 packet, packet no=9500 sendTime=1483527718432518
received 1024 packet, packet no=9500 sendTime=1483527718432518
received 1024 packet, packet no=9500 sendTime=1483527718432518
received 1024 packet, packet no=9500 sendTime=1483527718432518
received 1024 packet, packet no=9500 sendTime=1483527718432518
received 1024 packet, packet no=9500 sendTime=1483527718432518
received 1024 packet, packet no=9500 sendTime=1483527718432518
received 1024 packet, packet no=9500 sendTime=1483527718432518
received 1024 packet, packet no=9500 sendTime=1483527718432518
received 1024 packet, packet no=9500 sendTime=1483527718432518
received 1024 packet, packet no=9500 sendTime=1483527718432518
received 1024 packet, packet no=9500 sendTime=1483527718432518
received 1024 packet, packet no=9500 sendTime=1483527718432518
received 1024 packet, packet no=9500 sendTime=1483527718432518
received 1024 packet, packet no=9500 sendTime=1483527718432518
received 1024 packet, packet no=9500 sendTime=1483527718432518
received 1024 packet, packet no=9500 sendTime=1483527718432518
received 1024 packet, packet no=9500 sendTime=1483527718432518
received 1024 packet, packet no=9500 sendTime=1483527718432518
received 1024 packet, packet no=9500 sendTime=1483527718432518
received 1024 packet, packet no=9500 sendTime=1483527718432518
received 1024 packet, packet no=9500 sendTime=1483527718432518
received 1024 packet, packet no=9500 sendTime=1483527718432518
received 1024 packet, packet no=9500 sendTime=1483527718432518
received 1024 packet, packet no=9500 sendTime=1483527718432518
received 1024 packet, packet no=9500 sendTime=1483527718432518
received 1024 packet, packet no=9500 sendTime=1483527718432518
received 1024 packet, packet no=9500 sendTime=1483527718432518
received 1024 packet, packet no=10000 sendTime=1483527718439959
received 1024 packet, packet no=10000 sendTime=1483527718439959
received 1024 packet, packet no=10000 sendTime=1483527718439959
received 1024 packet, packet no=10000 sendTime=1483527718439959
received 1024 packet, packet no=10000 sendTime=1483527718439959
received 1024 packet, packet no=10000 sendTime=1483527718439959
received 1024 packet, packet no=10000 sendTime=1483527718439959
received 1024 packet, packet no=10000 sendTime=1483527718439959
received 1024 packet, packet no=10000 sendTime=1483527718439959
received 1024 packet, packet no=10000 sendTime=1483527718439959
received 1024 packet, packet no=10000 sendTime=1483527718439959
received 1024 packet, packet no=10000 sendTime=1483527718439959
received 1024 packet, packet no=10000 sendTime=1483527718439959
received 1024 packet, packet no=10000 sendTime=1483527718439959
received 1024 packet, packet no=10000 sendTime=1483527718439959
received 1024 packet, packet no=10000 sendTime=1483527718439959
received 1024 packet, packet no=10000 sendTime=1483527718439959
received 1024 packet, packet no=10000 sendTime=1483527718439959
received 1024 packet, packet no=10000 sendTime=1483527718439959
received 1024 packet, packet no=10000 sendTime=1483527718439959
received 1024 packet, packet no=10000 sendTime=1483527718439959
received 1024 packet, packet no=10000 sendTime=1483527718439959
received 1024 packet, packet no=10000 sendTime=1483527718439959
received 1024 packet, packet no=10000 sendTime=1483527718439959
received 1024 packet, packet no=10000 sendTime=1483527718439959
received 1024 packet, packet no=10000 sendTime=1483527718439959
received 1024 packet, packet no=10000 sendTime=1483527718439959
received 1024 packet, packet no=10000 sendTime=1483527718439959
received 1024 packet, packet no=10000 sendTime=1483527718439959
received 1024 packet, packet no=10000 sendTime=1483527718439959
received 1024 packet, packet no=10000 sendTime=1483527718439959
received 1024 packet, packet no=10000 sendTime=1483527718439959
received 1024 packet, packet no=10000 sendTime=1483527718439959
received 1024 packet, packet no=10000 sendTime=1483527718439959
received 1024 packet, packet no=10000 sendTime=1483527718439959
..............

If I didn't apply the zerocopy option, the last messages are lost.
In the log below, messages from 9514 to 9999 were lost.
..............
received 1024 packet, packet no=9500 sendTime=1483529380630926
received 1024 packet, packet no=9501 sendTime=1483529380638520
received 1024 packet, packet no=9502 sendTime=1483529380638532
received 1024 packet, packet no=9503 sendTime=1483529380638536
received 1024 packet, packet no=9504 sendTime=1483529380638539
received 1024 packet, packet no=9505 sendTime=1483529380638543
received 1024 packet, packet no=9506 sendTime=1483529380638547
received 1024 packet, packet no=9507 sendTime=1483529380638551
received 1024 packet, packet no=9508 sendTime=1483529380638554
received 1024 packet, packet no=9509 sendTime=1483529380638558
received 1024 packet, packet no=9510 sendTime=1483529380638562
received 1024 packet, packet no=9511 sendTime=1483529380638566
received 1024 packet, packet no=9512 sendTime=1483529380638570
received 1024 packet, packet no=9513 sendTime=1483529380638573

All problems work properly if I just add usleep(10) after sending 100 packets.
Since our test board has 4 ms clock tick, It will sleep 4ms even if I put the usleep(10).

Here is my publisher code applying zerocopy.
------------------------------------------------------------------------------------------------
int main (int argc, char *argv[])
{
    unsigned int     i = 0 ;
    unsigned long long time;
    cpu_set_t       CpuSet;
    pthread_t       Pid;
    int one = 1;
    TempMsg_Type    MsgBuf;
    int Len = 0;
    zmq_msg_t message;
    int subs = 0;
    unsigned int send_hwm = 0;

    char *buffer;

    memset((char *)&MsgBuf, 0, sizeof(TempMsg_Type));

    if ( argc != 2 )
    {
        printf("Usage : publish.<arch> <length#>\n");
        return 0;
    }

    // Prepare our context and publisher
    void *context = zmq_ctx_new ();
    assert (context);
    void *publisher = zmq_socket (context, ZMQ_PUB);
    assert(publisher);

    int rc = zmq_bind (publisher, "tcp://*:5556");
    if (rc != 0)
    {
        perror("zmq_bind() failed \n");
        return 0;
    }

    void *sync = zmq_socket(context, ZMQ_REP);
    rc = zmq_bind (sync, "tcp://*:5557");

    printf("Waiting for subscribers \n");
    while (subs < SUBSCRIBERS_EXPECTED)
    {
        /* synchronized */
        char *string = s_recv(sync);
        free(string);
        s_send(sync, "");
        subs++;
    }

    //set a hwm on publisher
    rc = zmq_setsockopt (publisher, ZMQ_SNDHWM, &send_hwm, sizeof (send_hwm));
    rc = zmq_setsockopt (publisher, ZMQ_XPUB_NODROP, &one, sizeof(one));
    // set a linger option
    rc = zmq_setsockopt (publisher, ZMQ_LINGER, &one, sizeof(one));
    rc = zmq_setsockopt (sync, ZMQ_LINGER, &one, sizeof(one));

    Len = atoi(argv[1]);
    printf("Allocating %d bytes of memory\n",Len);
    buffer = malloc(Len);
    if(buffer == NULL)
    {
        printf("Unable to allocate enough memory\n");
        zmq_close (publisher);
        zmq_close (sync);
        zmq_ctx_destroy (context);

        exit(-1);
    }

    printf("wait 10 seconds...\n");
    for (i = 0; i <10; i++)
    {
        printf(".");
        sleep(1);
    }

    startTime = getTimeUs();

    for (i = 1; i <= MSGNUM; i++) {

        MsgBuf.msgHdr.packetNo = i;
        time = getTimeUs();
        MsgBuf.msgHdr.sendTime = time;

        memcpy(buffer, (char *)&MsgBuf, Len);
        zmq_msg_init_data(&message, buffer, Len, NULL, NULL);

        zmq_msg_send(&message, publisher, 0);

        if ((i%100) == 0)
            usleep(10);

        printf("%d packet send, sendTime=%ld \n", MsgBuf.msgHdr.packetNo, MsgBuf.msgHdr.sendTime);
    }
    endTime = getTimeUs();
    diffTime = endTime - startTime;

    printf("Send %ld packets, start:%ld, end:%ld \n", i - 1, startTime, endTime);
    printf("Difftime : %ld us\n", diffTime);

    zmq_close (publisher);
    zmq_close (sync);
    zmq_msg_close(&message);
    zmq_ctx_destroy (context);

    while(1)
    {
        sleep(10);
    }

    return 0;
}

**Comment by bluca** [comment]
1) As the documentation says, using zero copy zmq_msg_init_data means that the library takes ownership of the buffer:

http://api.zeromq.org/4-2:zmq-msg-init-data

```
No copy of data shall be performed and ØMQ shall take ownership of the supplied buffer.
```

What's happening is that you are passing the same buffer to all messages, so obviously the content is the same. You need to malloc it inside the for loop, once for every message.

2) The lost messages are normal as HWM is probably getting overrun as your subscriber can't keep up with the publisher. Check the HWM documentation for information on how to fiddle with it:

http://api.zeromq.org/4-2:zmq-setsockopt

For more details on why the PUB-SUB was designed as lossy and how to handle see see the zguide:

http://zguide.zeromq.org/page:all#Getting-the-Message-Out

**Comment by bluca** [comment]
Could you please share the updated code? On a gist/pastebin maybe to avoid excessive clutter here

**Comment by clearday4** [comment]
I increased the message_pipe_granularity value of config.hpp from 256 to 1000 then the message loss disappeared.
And I set the send_hwm = 1000 and recv_hwm = 1,000,000.
Is this setting value valid?

**Comment by bluca** [comment]
No, you shouldn't change config.hpp. Can you share the code with the setting of HWM? Setting them to unlimited (remembering that this means unbounded memory usage) should be all you need

**Comment by clearday4** [comment]
With the above setting, one Subscriber successfully received 10000 packets.
However, I'm facing the problem registering 20 Subscribers.
Some Threads print out "error: 156384765 (Context was terminated)" and zmq_recv() was failed.
Finally, the following message printed out and the Subscriber ended.

Resource temporarily unavailable (src/signaler.cpp:282)
Aborted

The following code is hwm setting and send/received code.
I can't upload the test code to git cuase company's policy so I attach the part of test code.

**Comment by clearday4** [comment]
<pub.c>

#define SUBSCRIBERS_EXPECTED    20
#define MSGNUM          10000
int main (int argc, char *argv[])
{
    unsigned int     i = 0 ;
    unsigned long long time;
    cpu_set_t       CpuSet;
    int one = 1;
    TempMsg_Type    MsgBuf;
    int Len = 0;
    zmq_msg_t message;
    int subs = 0;
    int index = 0;
    unsigned int    sendCnt = 0;

    TestBuf_Type    *buffer;

    CPU_SET(1, &CpuSet);
    sched_setaffinity( 0, sizeof(cpu_set_t), &CpuSet);

    memset((char *)&MsgBuf, 0, sizeof(TempMsg_Type));

    if ( argc != 2 )
    {
        printf("Usage : publish.<arch> <length#>\n");
        return 0;
    }

    // Prepare our context and publisher
    void *context = zmq_ctx_new ();
    assert (context);
    void *publisher = zmq_socket (context, ZMQ_PUB);
    assert(publisher);

    int rc = zmq_bind (publisher, "tcp://*:5556");
    if (rc != 0)
    {
        perror("zmq_bind() failed \n");
        return 0;
    }

    void *sync = zmq_socket(context, ZMQ_REP);
    rc = zmq_bind (sync, "tcp://*:5557");

    printf("Waiting for subscribers \n");
    while (subs < SUBSCRIBERS_EXPECTED)
    {
        /* synchronized */
        char *string = s_recv(sync);
        free(string);
        s_send(sync, "");
        subs++;
    }

    //set a hwm on publisher
    unsigned int send_hwm = 1000;
    rc = zmq_setsockopt (publisher, ZMQ_SNDHWM, &send_hwm, sizeof (send_hwm));
    rc = zmq_setsockopt (publisher, ZMQ_XPUB_NODROP, &one, sizeof(one));
    // set a linger option
    rc = zmq_setsockopt (publisher, ZMQ_LINGER, &one, sizeof(one));
    rc = zmq_setsockopt (sync, ZMQ_LINGER, &one, sizeof(one));

    buffer =  (TestBuf_Type *)malloc(sizeof(TestBuf_Type) * 10000);
    if(buffer == NULL)
    {
        printf("Unable to allocate enough memory\n");
        zmq_close (publisher);
        zmq_close (sync);
        zmq_ctx_destroy (context);

        exit(-1);
    }

    printf("wait 15 seconds...\n");
    for (i = 0; i <15; i++)
    {
        printf(".");
        sleep(1);
    }

    startTime = getTimeUs();

    for (i = 1; i <= MSGNUM; i++) {

        MsgBuf.msgHdr.packetNo = i;
        time = getTimeUs();
        MsgBuf.msgHdr.sendTime = time;

        index = i-1;

        memcpy(buffer[index].buf, (char *)&MsgBuf, Len);
        zmq_msg_init_data(&message, buffer[index].buf, Len, NULL, NULL);

        do {
            rc = zmq_msg_send(&message, publisher, ZMQ_DONTWAIT);
            if (rc == -1)
            {
                if (errno == EAGAIN)
                {
        //          printf("message reaches HWM \n");
                }
            }
            else
            {
                sendCnt++;
                break;
            }
        } while (rc < 0);

//      if ((i%100) == 0)
            usleep(10);
//      printf("%d packet send, sendTime=%ld \n", MsgBuf.msgHdr.packetNo, MsgBuf.msgHdr.sendTime);
    }
    endTime = getTimeUs();
    diffTime = endTime - startTime;

    printf("Send %ld packets, start:%ld, end:%ld \n", sendCnt, startTime, endTime);
    printf("Difftime : %ld us\n", diffTime);

    zmq_close (publisher);
    zmq_close (sync);
    zmq_msg_close(&message);
    zmq_ctx_destroy (context);

    return 0;
}

<sub.c>
void subscriber(void *pArg)
{
    int rc;
    int one = 1;
    unsigned int    recvCnt = 0;
    unsigned long long getTime = 0,calTime = 0;

    TempMsg_Type *pMsg;
    // Prepare our context and subscriber
    void *subscriber;
    zmq_msg_t   message;

    subscriber = zmq_socket (context, ZMQ_SUB);
    assert(subscriber);

    rc = zmq_setsockopt (subscriber, ZMQ_LINGER, &one, sizeof(one));

    int *pIndex = pArg;
    int Index = *pIndex;

    rc = zmq_setsockopt(subscriber, ZMQ_SUBSCRIBE, NULL, 0);

    rc = zmq_connect (subscriber, "tcp://10.10.10.10:5556");
    if (rc != 0)
    {
        printf("[%d] zmq_connect() failed\n", Index);
        zmq_close (subscriber);
        return;
    }

    //set a hwm on publisher
    unsigned int recv_hwm = 0;
   // unsigned int recv_hwm = 1000000;
    rc = zmq_setsockopt (subscriber, ZMQ_RCVHWM, &recv_hwm, sizeof (recv_hwm));

    sleep(1);
    void *sync = zmq_socket(context, ZMQ_REQ);
    rc = zmq_setsockopt (sync, ZMQ_LINGER, &one, sizeof(one));
    rc = zmq_connect(sync, "tcp:/10.10.10.10:5557");
    if (rc != 0)
    {
        printf("[%d] zmq_connect() sync failed\n", Index);
        zmq_close (sync);
        return;
    }

    printf("Send sync messages \n");
    s_send(sync, "");
    char *string = s_recv(sync);
    free(string);

    printf("Start receiving ... \n");
    zmq_msg_init(&message);

    while (1) {

        rc = zmq_recv (subscriber, buffer, MAX_MSG_LENGTH, ZMQ_DONTWAIT);
        if (rc < 0)
        {
            if (errno == EAGAIN)
            {
                continue;
            }
            printf("[%d] error : %d (%s)\n", Index, errno, zmq_strerror(errno));
        }
        pMsg = (TempMsg_Type *)&buffer;

        recvCnt++;

        if (recvCnt == 1)
        {
            startTime = getTimeUs();
        }

        //printf("%d] received %d packet, packet no=%d sendTime=%ld\n", recvCnt, rc, pMsg->msgHdr.packetNo, pMsg->msgHdr.sendTime);

        if (recvCnt == MSGNUM)
        {
            endTime = getTimeUs();
            break;
        }

        getTime = getTimeUs();
        calTime = getTime - pMsg->msgHdr.sendTime;
        //      printf("%d) getTime=%ld, calTime=%ld \n", recvCnt, getTime, calTime);

        if ((saveTime[Index].maxTime == 0) || (calTime > saveTime[Index].maxTime))
        {
            saveTime[Index].maxTime = calTime;
        }

        if ((saveTime[Index].minTime == 0) || (calTime < saveTime[Index].minTime))
        {
            saveTime[Index].minTime = calTime;
        }

        saveTime[Index].accTime = saveTime[Index].accTime + calTime;
    }

    diffTime = endTime - startTime;

    printf("\n=========================[%d] Sub Result ======================== \n", Index);
    printf("[%d] %d packet rcvd, startTime:%ld, endTime:%ld \n", Index, recvCnt, startTime, endTime);
    printf("[%d] diffTime:%ld \n", Index, diffTime);
    printf("[%d] Latency max:%ld, min:%ld, average:%ld \n", Index, saveTime[Index].maxTime, saveTime[Index].minTime,  (saveTime[Index].accTime/MSGNUM));
    printf("=================================================================== \n");

    zmq_close (subscriber);
    zmq_close (sync);
    zmq_msg_close(&message);
    zmq_ctx_destroy (context);
    return ;
}

**Comment by bluca** [comment]
4.1.3 is quite old, please use 4.2.1 or the master branch of this repository. Also those code snippets don't compile

**Comment by yunfengliu2** [comment]
maybe you should check that the change to HWM settings must occur after socket creation and before bind()/connect() in your <pub.c> ,see issue #1416. @clearday4

**Comment by bluca** [comment]
No further update in a year, assuming fixed when moving from 4.1.x to 4.2.x, closing

**Comment by lqxandxl** [comment]
@clearday4  
I use your way change message_pipe_granularity to 1000，and it`s useful,but i don`t know why it`s useful.

i use zeromq4.2.5 and openpgm in windows, use zmq_pub, sending 90000 msgs / s, 200B one message.
sub receive msgs and then speed come down to 0,and never receive msg again.

seems like in fk.cpp the ative value is 0,and always return -1,when i use function recv.

another thing is  in good computer,things do not get wrong. but in bad computer, it has problem.

but i don`t know why ,  my  rcvhwm value set 1024*1024*1024
